### ISSUE TITLE
iron-icon accessibility report

### ISSUE THREAD

**Issue by lpalmaro** [PolymerElements/iron-icon#8]
Here are the accessibility developer tools audit results:

[Warning] Images should have an alt attribute (1)[Warning] Text elements should have a reasonable contrast ratio (1)Not applicable tests (9)Passing tests (4)

The only issue here is that the icons don't have labels, so when using a screen reader, you don't receive info about what the icons are for/represent. I know this is simply a demo page, so it's not a concern at all. Would be good to make sure it's very easy for developers to add alt text appropriately though.

**Comment by alice** [comment]
Suggestion: use `alt=""` on internal `img` element.
Suggestion: don't apply `role` since in the vast majority of cases these will be presentational only. The labelling needs to come from somewhere else - e.g. in the case of an icon button, either the button text or a label on the button.

**Comment by cdata** [comment]
> The only issue here is that the icons don't have labels

In previous versions, the icons attempted to automatically select a label for themselves, which initially amounted to picking the icon name as the label (if available).

This starts to get in the way as soon as icons are incorporated into other widgets (ones that may or may not have an alternative label source). With this in mind, it seemed better to treat icons like img elements and expect that they will be given a label by an author with special knowledge of the context.

What are your thoughts here? Would it be satisfactory to apply labels to the demo "manually"?

**Comment by alice** [comment]
My suggestion is to apply empty alt text to the img element and not add the `img` role to the shadow host, since in the majority of cases these icons won't be "acting as" an image, but as presentation. For example, in the case of a button with a 'tick' icon and 'Ok' text, the icon should be completely hidden from AT as it's redundant with the text. In the case of a piece of UI where the icon provides the only clue as to its purpose, that piece of UI (e.g. the `paper-icon-button`) should be the place where the text alternative is attached, rather than the icon.

**Comment by lpalmaro** [comment]
Ping on progress? Thanks!

**Comment by alice** [comment]
On the alt text issue: I have just had a chat with @robdodson who is using an `<iron-icon>` inside an `<a>`. In this case, it would be really nice to be able to apply some kind of text alternative on the icon itself, so that the `<a>` can pick it up as text content. So, basically the reverse of what I said previously :)

**Comment by jannakha** [comment]
+1
what's the progress?

**Comment by robdodson** [comment]
Looking into this

**Comment by abdonrd** [comment]
News about this?

**Comment by robdodson** [comment]
here's my thinking, @alice please correct me if i'm wrong.
- There's no need to add a role to the host because there's an img tag inside of iron-icon and the accessibility tree will pick that up
- Typically iron-icon is presentational so shouldn't auto apply any sort of label or alt text to the underlying img
- Inside of paper-icon-button, the label should go on the button and not the iron-icon so no need to change anything there
- If inside an `<a>` the developer could add aria-label to the anchor, or add aria-label to the iron-icon. Alternatively iron-icon could support an `alt` attribute which would be bound to the `alt` attribute of the img in the shadow dom. Unfortunately that img is created with JavaScript so there would need to be an observer that watches for changes to `alt` on the host and manually updates the img.

**Comment by notwaldorf** [comment]
Small nit: if the `iron-icon` gets an `alt` attribute, it should _not_ bind it to the `img`'s `alt` (exposing implementation details), but it should instead just update its host's `aria-label`, like we do for `paper-icon-button`: https://github.com/PolymerElements/paper-icon-button/blob/master/paper-icon-button.html#L163

**Comment by robdodson** [comment]
Maybe also worth mentioning. The element only uses an img if the icon is based on a raster format. It may also use iron-iconset-svg internally. So I guess i'd lean toward using aria-label as well

**Comment by alice** [comment]
I would have thought `aria-label` was more "leaky" since it'd sprout on the shadow host. If you support `alt`, you could _implement_ support for it as appropriate inside of shadows - either `alt` on an `<img>` element or `aria-label` on an alternative format. But honestly I'd be fine with sprouting `aria-label` as well, as long as you set `role=img` at the same time (imagine seeing a standalone `iron-icon` on the page).

¯\_(ツ)_/¯

**Comment by patrinhani-ciandt** [comment]
@robdodson and @notwaldorf, I sent a [Pull Request](https://github.com/PolymerElements/iron-icon/pull/77). Please, take a look if it makes sense.

**Comment by erikkroes** [comment]
@alice @robdodson My shot at a fix for this: https://github.com/PolymerElements/iron-icon/pull/101

No `alt` or empty `alt` makes the host `aria-hidden`.
If there is an `alt`, an `aria-label` is added, and `role="img"`.

So you either set an alt explicitly, or it is ignored completely by screen readers.

**Comment by phidias51** [comment]
I guess I'm a little confused about this. Are you supposed to set "alt" or "aria-label"?  I've tried both, and neither of them work. I'm using iron-icon v1.0.13.

**Comment by ghost** [comment]
I can confirm @phidias51 in that neither alt or aria-label in the iron-icon element are passed down to the actual img element. This is super annoying since I can't get a 100% accessibility score with Lighthouse until this issue is resolved !!(>_<)!!

**Comment by robdodson** [comment]
I'd suggest iron-icon follow the same approach as the Angular Material icons.
https://material.angular.io/components/icon/overview#accessibility

By default, put `aria-hidden=true` on iron-icon and if you're using it inside of a button or something like that, then make sure the button has an accessible label. If it's an indicator then add a screen reader only span with the indicator text.

**Comment by phidias51** [comment]
What's also a bit confusing is how tooltips should work in a mobile setting versus a desktop. If I hover over an icon on a desktop, then I expect to see a tooltip pop up. If I'm on a mobile, there is no hover event, so I would the user trigger a tooltip?

I also don't know what the aria-hidden=true setting will do.  Will this hide the description all of the time, and only be "visible" to screenreaders.  What's needed is a universal way to access a description for an icon, regardless of the device, regardless of the browser.

**Comment by phidias51** [comment]
Although, it doesn't address the question of how tooltips should be triggered in a mobile setting (or by screenreaders), I've found that the **paper-tooltip** solves part of this problem.

**Comment by robdodson** [comment]
aria-hidden=true or role=presentation means a screen reader will completely skip the icon. Instead, the labeling should be on the button itself, and the icon ignored.

**Comment by erikkroes** [comment]
@robdodson So the "official" solution right now is `aria-hidden` or a screen reader only span? Then this issue can be closed right?
